### PR TITLE
Chat

### PR DESCRIPTION
**Pull request by Aliaksandra-Leanovich**

### CODE REVIEW

File: spa-react/src/components/Chat/Chat.tsx [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  TextSC,
+} from "./style";
+
+export const Chat = () => {
+  const { reset } = useForm();
+  const { t } = useTranslation();
+  const dispatch = useAppDispatch();
+
+  const chatId = useAppSelector(getChatId);
+  const messages = useAppSelector(getMessages);
```
Comment (rnavarych) [inline spa-react/src/components/Chat/Chat.tsx]: move to hook

Comment (rnavarych) [reply spa-react/src/components/Chat/Chat.tsx]: I mean all functions and data. We don't need all this logic and data in component

File: spa-react/src/components/Chat/Chat.tsx [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+  const [sent, setSent] = useState<IMessage[]>();
+  const [receive, setReceive] = useState<IMessage[]>();
+
+  const [edit, setEdit] = useState<boolean>(false);
+  const [messageId, setMessageId] = useState<string>("");
+
+  useEffect(() => {
+    if (userEmail) {
+      onSnapshot(collection(db, "messages"), (querySnapshot) => {
```
Comment (rnavarych) [inline spa-react/src/components/Chat/Chat.tsx]: move

File: spa-react/src/components/Chat/Chat.tsx [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        setSent(filteredSent);
+        setReceive(filteredReceive);
+      });
+    }
+  }, [recieverEmail, userEmail, dispatch]);
+
+  useEffect(() => {
+    dispatch(resetAllMessages());
+
+    sent?.map((item) => {
```
Comment (rnavarych) [inline spa-react/src/components/Chat/Chat.tsx]: looks strange. Why do you need here useEffect for sending of messages?!

File: spa-react/src/components/Chat/Chat.tsx [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    const diffrentUser = message.senderName !== previous.senderName;
+    if (diffrentUser) return true;
+
+    const hasBeenaWhile = message.date.seconds - previous.date.seconds > 60;
+    return hasBeenaWhile;
+  };
+
+  const handleDelete = async (message: IMessage) => {
+    try {
+      await deleteDoc(doc(db, "messages", message.id));
```
Comment (rnavarych) [inline spa-react/src/components/Chat/Chat.tsx]: you can use constants for keys of DB

File: spa-react/src/components/Chat/Chat.tsx [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  const handleEdit = async (message: IMessage) => {
+    dispatch(setValue(message.message));
+
+    reset((formValues) => ({
+      ...formValues,
+      message: message.message,
+    }));
+
+    setEdit(true);
+    setMessageId(message.id);
```
Comment (rnavarych) [inline spa-react/src/components/Chat/Chat.tsx]: it looks like excess logic? Why do you need it?

File: spa-react/src/components/InputChat/InputChat.tsx [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        console.log("error", err);
+      }
+    } else {
+      let setId = uuid();
+
+      if (message) {
+        await setDoc(doc(db, "messages", setId), {
+          id: setId,
+          message: message,
+          date: new Date(),
```
Comment (rnavarych) [inline spa-react/src/components/InputChat/InputChat.tsx]: UTC?

File: spa-react/src/components/InputChat/InputChat.tsx [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+          senderEmail: userEmail,
+          senderName: userName,
+          receivers: [receiverEmail],
```
Comment (rnavarych) [inline spa-react/src/components/InputChat/InputChat.tsx]: Why do you need it?

File: spa-react/src/components/InputChat/InputChat.tsx [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    reset((formValues) => ({
+      ...formValues,
+      message: "",
+    }));
+
+    dispatch(setValue(""));
+  };
+
+  useEffect(() => {
+    reset({
```
Comment (rnavarych) [inline spa-react/src/components/InputChat/InputChat.tsx:43]: Are you sure that you need useEffect for that?

File: spa-react/src/components/Message/Message.tsx [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        <TextSC>{message.message}</TextSC>
+      </InfoSC>
+
+      {message.senderEmail === userEmail && (
+        <ButtonSC onClick={() => handleDelete(message)}>
+          <Delete />
+        </ButtonSC>
+      )}
+      {message.senderEmail === userEmail && (
+        <ButtonSC onClick={() => handleEdit(message)}>
```
Comment (rnavarych) [inline spa-react/src/components/Message/Message.tsx]: this is old problem! Why do you add function for onClick event like that!?! You will create a functions every render!!!!

File: spa-react/src/components/Message/types.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  };
+}
+
+export interface IMessage {
+  date: { seconds: number; nanoseconds: number };
+  id: string;
+  message: string;
+  senderEmail: string;
+  senderName: string;
+  receivers: [];
```
Comment (rnavarych) [inline spa-react/src/components/Message/types.ts]: rename

File: spa-react/src/components/Message/types.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    date: { seconds: number; nanoseconds: number };
+  };
+}
+
+export interface IMessage {
+  date: { seconds: number; nanoseconds: number };
+  id: string;
+  message: string;
+  senderEmail: string;
+  senderName: string;
```
Comment (rnavarych) [inline spa-react/src/components/Message/types.ts]: why you don't want to use object?!

File: spa-react/src/components/Navigation/Navigation.tsx [hunk 12]
```diff
@@ -71,9 +70,10 @@ export const Navigation = () => {
       <>
         {isAuthorized ? (
           <>
-            <Link to={routes.USERS} variant={LinkVariants.primaryWhiteSmall}>
+            <LinkCustomSC to={routes.CHATS}>{t("nav.chat")}</LinkCustomSC>
+            {/* <Link to={routes.USERS} variant={LinkVariants.primaryWhiteSmall}>
```
Comment (rnavarych) [inline spa-react/src/components/Navigation/Navigation.tsx]: Why do you add this comments?

File: spa-react/src/components/Sidebar/Sidebar.tsx [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+export const Sidebar = () => {
+  const [users, setUsers] = useState<IUser[]>();
+
+  const userId = useAppSelector(getUserId);
+  const userEmail = useAppSelector(getUserEmail);
+
+  const dispatch = useAppDispatch();
+
+  useEffect(() => {
+    onSnapshot(collection(db, "users"), (querySnapshot) => {
```
Comment (rnavarych) [inline spa-react/src/components/Sidebar/Sidebar.tsx]: move

File: spa-react/src/components/Sidebar/Sidebar.tsx [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+      const users: any[] = [];
+      querySnapshot.forEach((doc) => {
+        users.push(doc.data());
+      });
+      setUsers(users);
+    });
+  }, []);
+
+  const handleSelect = async (id: string) => {
+    const combinedId = userId + id;
```
Comment (rnavarych) [inline spa-react/src/components/Sidebar/Sidebar.tsx]: looks really strange. Why?

**Review by rnavarych — CHANGES_REQUESTED** [review]

File: spa-react/src/components/Chat/Chat.tsx (hunk 3 above)
Comment (Aliaksandra-Leanovich) [reply spa-react/src/components/Chat/Chat.tsx]: we put messages which we sent to firestore , and then take them from firestore and put to the redux. So this useEffect take messages which we sent and which were sent to us

File: spa-react/src/components/Chat/Chat.tsx (hunk 5 above)
Comment (Aliaksandra-Leanovich) [reply spa-react/src/components/Chat/Chat.tsx]: We put message's id to the state, and take it when send updated message to the firestore. So we need it to find message which we need to update by id

File: spa-react/src/components/InputChat/InputChat.tsx (hunk 8 above)
Comment (Aliaksandra-Leanovich) [reply spa-react/src/components/InputChat/InputChat.tsx:43]: I'm sure. Without this useEffect it wont put value which we need to the input . I tried to remove and rewrite code  today again, however, it didnt work

File: spa-react/src/components/InputChat/InputChat.tsx (hunk 7 above)
Comment (Aliaksandra-Leanovich) [reply spa-react/src/components/InputChat/InputChat.tsx]: we need to sort messages by whom they need to be received to . We can have chats with 2 and more users in future (for example)

File: spa-react/src/components/Message/types.ts (hunk 10 above)
Comment (Aliaksandra-Leanovich) [reply spa-react/src/components/Message/types.ts]: we need to sort messages by whom they need to be received to . We can have chats with 2 and more users in future (for example). So we are the sender and receivers - who will receive our messages

**Pull request merged**